tsdb/shard.go

```diff
@@ -813,6 +1065,129 @@ func u64tob(v uint64) []byte {
+	v = append(v, key...)
+	v = append(v, data...)
+	return v
+}
+
+// unmarshalWALEntry decodes a WAL entry into it's separate parts.
+// Returned byte slices point to the original slice.
+func unmarshalWALEntry(v []byte) (key []byte, timestamp int64, data []byte) {
+	keyLen := binary.BigEndian.Uint32(v[8:12])
+	key = v[12 : 12+keyLen]
+	timestamp = int64(binary.BigEndian.Uint64(v[0:8]))
+	data = v[12+keyLen:]
+	return
+}
+
+// marshalCacheEntry encodes the timestamp and data to a single byte slice.
+//
+// The format of the byte slice is:
+//
+//     uint64 timestamp
+//     []byte data
+//
+func marshalCacheEntry(timestamp int64, data []byte) []byte {
+	buf := make([]byte, 8, 8+len(data))
+	binary.BigEndian.PutUint64(buf[0:8], uint64(timestamp))
+	return append(buf, data...)
+}
+
+// unmarshalCacheEntry returns the timestamp and data from an encoded byte slice.
+func unmarshalCacheEntry(buf []byte) (timestamp int64, data []byte) {
+	timestamp = int64(binary.BigEndian.Uint64(buf[0:8]))
+	data = buf[8:]
+	return
+}
+
+// byteSlices represents a sortable slice of byte slices.
+type byteSlices [][]byte
+
+func (a byteSlices) Len() int           { return len(a) }
+func (a byteSlices) Less(i, j int) bool { return bytes.Compare(a[i], a[j]) == -1 }
+func (a byteSlices) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+// shardCursor provides ordered iteration across a Bolt bucket and shard cache.
+type shardCursor struct {
+	// Bolt cursor and readahead buffer.
+	cursor *bolt.Cursor
+	buf    struct {
+		key, value []byte
+	}
+
+	// Cache and current cache index.
+	cache [][]byte
+	index int
+}
+
+// Seek moves the cursor to a position and returns the closest key/value pair.
+func (sc *shardCursor) Seek(seek []byte) (key, value []byte) {
+	// Seek bolt cursor.
+	if sc.cursor != nil {
+		sc.buf.key, sc.buf.value = sc.cursor.Seek(seek)
+	}
+
+	// Seek cache index.
+	sc.index = sort.Search(len(sc.cache), func(i int) bool {
+		return bytes.Compare(sc.cache[i][0:8], seek) != -1
+	})
+
+	return sc.read()
+}
+
+// Next returns the next key/value pair from the cursor.
+func (sc *shardCursor) Next() (key, value []byte) {
+	// Read next bolt key/value if not bufferred.
+	if sc.buf.key == nil && sc.cursor != nil {
+		sc.buf.key, sc.buf.value = sc.cursor.Next()
+	}
+
+	return sc.read()
+}
+
+// read returns the next key/value in the cursor buffer or cache.
+func (sc *shardCursor) read() (key, value []byte) {
+	// If neither a buffer or cache exists then return nil.
+	if sc.buf.key == nil && sc.index >= len(sc.cache) {
+		return nil, nil
+	}
+
+	// Use the buffer if it exists and there's no cache or if it is lower than the cache.
+	if sc.buf.key != nil && (sc.index >= len(sc.cache) || bytes.Compare(sc.buf.key, sc.cache[sc.index][0:8]) == -1) {
+		key, value = sc.buf.key, sc.buf.value
+		sc.buf.key, sc.buf.value = nil, nil
+		return
+	}
+
+	// Otherwise read from the cache.
+	// Continue skipping ahead through duplicate keys in the cache list.
```

Comment:
How could there be duplicate keys? Points for the series written for the same timestamp?